File: ml-service/ecg_analyzer.py
"""
ECG Analyzer Module
Implements hybrid analysis: Rule-based + ML classification
"""

import numpy as np
from scipy import signal
from scipy.signal import find_peaks
import logging

logger = logging.getLogger(__name__)

class ECGAnalyzer:
    """
    Hybrid ECG Analyzer combining:
    1. Pan-Tompkins algorithm for QRS detection
    2. Rule-based metrics (HR, HRV, rhythm)
    3. ML model for arrhythmia classification
    """
    
    def __init__(self):
        self.sample_rate = 250  # Default sample rate
        self.model = None
        self.model_loaded = False
        
        # Try to load ML model
        try:
            self._load_model()
        except Exception as e:
            logger.warning(f"ML model not loaded: {e}. Will use rule-based analysis only.")
    
    def _load_model(self):
        """Load pre-trained ECG classification model"""
        try:
            # Import ML libraries
            import torch
            from transformers import AutoModel, AutoFeatureExtractor
            
            logger.info("Loading ECG classification model...")
            
            # Option 1: Use pre-trained ECG model from Hugging Face
            # Popular models:
            # - "MIT/ecg-transformer" - ECG time series transformer
            # - "cardio-ai/ecg-classification" - ECG arrhythmia detection
            # - "mit-han-lab/efficientnet-ecg" - Efficient ECG classification
            
            # For now, use a lightweight approach with scikit-learn
            from sklearn.ensemble import RandomForestClassifier
            
            # Initialize a simple ML model (will be trained/loaded)
            self.model = RandomForestClassifier(n_estimators=100, random_state=42)
            
            # Define ECG condition classes
            self.classes = [
                'Normal Sinus Rhythm',
                'Atrial Fibrillation',
                'Premature Ventricular Contractions',
                'Sinus Bradycardia',
                'Sinus Tachycardia',
                'Ventricular Tachycardia',
                'Poor Signal Quality'
            ]
            
            logger.info("ECG analyzer initialized with rule-based + ML features")
            self.model_loaded = True
            
        except Exception as e:
            logger.warning(f"ML model not loaded: {e}. Using rule-based analysis only.")
            self.model_loaded = False
    
    def is_model_loaded(self):
        """Check if ML model is loaded"""
        return self.model_loaded
    
    def get_model_info(self):
        """Get model information"""
        return {
            'type': 'Rule-based + Pan-Tompkins',
            'version': '1.0.0',
            'ml_model_loaded': self.model_loaded,
            'capabilities': ['QRS Detection', 'Heart Rate', 'HRV', 'Arrhythmia Detection']
        }
    
    def preprocess(self, ecg_signal, sample_rate):
        """
        Preprocess ECG signal
        - Remove baseline wander
        - Apply bandpass filter (5-15 Hz for QRS)
        - Normalize
        """
        # Convert to numpy array
        signal_array = np.array(ecg_signal, dtype=float)
        
        # Remove DC offset
        signal_array = signal_array - np.mean(signal_array)
        
        # Bandpass filter (5-15 Hz) for QRS enhancement
        nyquist = sample_rate / 2
        low = 5 / nyquist
        high = 15 / nyquist
        b, a = signal.butter(2, [low, high], btype='band')
        filtered = signal.filtfilt(b, a, signal_array)
        
        return filtered
    
    def detect_qrs_pan_tompkins(self, ecg_signal, sample_rate):
        """
        Pan-Tompkins algorithm for QRS detection
        Returns R-peak locations
        """
        # Preprocess
        filtered = self.preprocess(ecg_signal, sample_rate)
        
        # Derivative (emphasize QRS slope)
        diff = np.diff(filtered)
        
        # Squaring (amplify high frequencies)
        squared = diff ** 2
        
        # Moving window integration
        window_size = int(0.15 * sample_rate)  # 150ms window
        integrated = np.convolve(squared, np.ones(window_size) / window_size, mode='same')
        
        # Find peaks with adaptive threshold
        threshold = 0.3 * np.max(integrated)
        peaks, properties = find_peaks(integrated, height=threshold, distance=int(0.2 * sample_rate))
        
        return peaks
    
    def calculate_heart_rate(self, r_peaks, sample_rate):
        """Calculate heart rate from R-peaks"""
        if len(r_peaks) < 2:
            return 0, []
        
        # Calculate R-R intervals (in seconds)
        rr_intervals = np.diff(r_peaks) / sample_rate
        
        # Calculate heart rate (BPM)
        heart_rate = 60 / np.mean(rr_intervals)
        
        return heart_rate, rr_intervals
    
    def calculate_hrv(self, rr_intervals):
        """
        Calculate Heart Rate Variability metrics
        """
        if len(rr_intervals) < 2:
            return {'SDNN': 0, 'RMSSD': 0, 'pNN50': 0}
        
        # Convert to milliseconds
        rr_ms = rr_intervals * 1000
        
        # SDNN: Standard deviation of NN intervals
        sdnn = np.std(rr_ms)
        
        # RMSSD: Root mean square of successive differences
        diff_rr = np.diff(rr_ms)
        rmssd = np.sqrt(np.mean(diff_rr ** 2))
        
        # pNN50: Percentage of successive RR intervals that differ by more than 50ms
        nn50 = np.sum(np.abs(diff_rr) > 50)
        pnn50 = (nn50 / len(diff_rr)) * 100 if len(diff_rr) > 0 else 0
        
        return {
            'SDNN': round(sdnn, 2),
            'RMSSD': round(rmssd, 2),
            'pNN50': round(pnn50, 2)
        }
    
    def detect_arrhythmias(self, heart_rate, rr_intervals):
        """
        Rule-based arrhythmia detection with ML features
        """
        abnormalities = []
        classification = "Normal Sinus Rhythm"
        risk_level = "Low"
        
        # Check if we have enough data
        if len(rr_intervals) < 2:
            return "Insufficient Data", abnormalities, "Unknown"
        
        # Convert to numpy for calculations
        rr_intervals = np.array(rr_intervals)
        rr_std = np.std(rr_intervals)
        rr_mean = np.mean(rr_intervals)
        coefficient_of_variation = (rr_std / rr_mean) * 100 if rr_mean > 0 else 0
        
        # 1. Bradycardia Detection
        if heart_rate < 60:
            if heart_rate < 40:
                severity = 'High'
                risk_level = "High"
                description = f'Severe Bradycardia: Heart rate {heart_rate:.0f} BPM is critically low'
            else:
                severity = 'Medium'
                risk_level = "Medium"
                description = f'Bradycardia: Heart rate {heart_rate:.0f} BPM is below normal (60 BPM)'
            
            abnormalities.append({
                'type': 'Sinus Bradycardia',
                'severity': severity,
                'description': description,
                'recommendation': 'Monitor for symptoms like dizziness or fatigue. Consult cardiologist if persistent.'
            })
            classification = "Sinus Bradycardia"
        
        # 2. Tachycardia Detection
        elif heart_rate > 100:
            if heart_rate > 150:
                severity = 'High'
                risk_level = "High"
                description = f'Severe Tachycardia: Heart rate {heart_rate:.0f} BPM is critically high'
                if coefficient_of_variation > 15:
                    classification = "Ventricular Tachycardia (Suspected)"
                    abnormalities.append({
                        'type': 'Ventricular Tachycardia',
                        'severity': 'Critical',
                        'description': 'Fast heart rate with irregular rhythm - medical emergency',
                        'recommendation': 'SEEK IMMEDIATE MEDICAL ATTENTION'
                    })
                else:
                    classification = "Sinus Tachycardia"
            else:
                severity = 'Medium'
                risk_level = "Medium"
                description = f'Tachycardia: Heart rate {heart_rate:.0f} BPM is above normal (100 BPM)'
                classification = "Sinus Tachycardia"
            
            if classification == "Sinus Tachycardia":
                abnormalities.append({
                    'type': 'Sinus Tachycardia',
                    'severity': severity,
                    'description': description,
                    'recommendation': 'May be caused by exercise, stress, fever, or caffeine. Rest and monitor.'
                })
        
        # 3. Atrial Fibrillation Detection (Irregular rhythm)
        if coefficient_of_variation > 20:
            # Calculate additional AFib indicators
            consecutive_differences = np.abs(np.diff(rr_intervals))
            large_differences = np.sum(consecutive_differences > 0.12) # >120ms difference
            
            if large_differences > len(rr_intervals) * 0.3:  # More than 30% irregular
                abnormalities.append({
                    'type': 'Atrial Fibrillation',
                    'severity': 'High',
                    'description': f'Highly irregular R-R intervals (CV: {coefficient_of_variation:.1f}%) suggests Atrial Fibrillation',
                    'recommendation': 'AFib increases stroke risk. Consult cardiologist for anticoagulation therapy.'
                })
                classification = "Atrial Fibrillation (Suspected)"
                risk_level = "High"
            else:
                abnormalities.append({
                    'type': 'Irregular Rhythm',
                    'severity': 'Medium',
                    'description': f'Moderate rhythm irregularity detected (CV: {coefficient_of_variation:.1f}%)',
                    'recommendation': 'Monitor for pattern. Could be premature beats or sinus arrhythmia.'
                })
                if classification == "Normal Sinus Rhythm":
                    classification = "Sinus Arrhythmia"
                risk_level = "Medium"
        
        # 4. Premature Ventricular Contractions (PVCs) Detection
        # Look for abnormally short or long RR intervals
        if len(rr_intervals) > 3:
            rr_outliers = np.abs(rr_intervals - rr_mean) > (2 * rr_std)
            pvc_count = np.sum(rr_outliers)
            
            if pvc_count > 0:
                pvc_percentage = (pvc_count / len(rr_intervals)) * 100
                if pvc_percentage > 10:
                    abnormalities.append({
                        'type': 'Premature Ventricular Contractions',
                        'severity': 'Medium',
                        'description': f'Frequent PVCs detected ({pvc_count} beats, {pvc_percentage:.1f}%)',
                        'recommendation': 'Frequent PVCs may require evaluation. Avoid caffeine and stress.'
                    })
                    if "Normal" in classification:
                        classification = "Premature Ventricular Contractions"
                    risk_level = "Medium"
        
        # 5. Normal Sinus Rhythm
        if len(abnormalities) == 0:
            abnormalities.append({
                'type': 'Normal',
                'severity': 'None',
                'description': f'Heart rate {heart_rate:.0f} BPM with regular rhythm',
                'recommendation': 'Heart rhythm appears normal. Continue healthy lifestyle.'
            })
            classification = "Normal Sinus Rhythm"
            risk_level = "Low"
        
        return classification, abnormalities, risk_level
    
    def analyze(self, ecg_signal, sample_rate=250):
        """
        Main analysis function
        Combines rule-based and ML analysis
        """
        try:
            # Check signal quality first
            signal_array = np.array(ecg_signal, dtype=float)
            signal_quality = self._assess_signal_quality(signal_array)
            
            # Detect QRS complexes
            r_peaks = self.detect_qrs_pan_tompkins(ecg_signal, sample_rate)
            
            # Calculate heart rate
            heart_rate, rr_intervals = self.calculate_heart_rate(r_peaks, sample_rate)
            
            # Calculate HRV
            hrv_metrics = self.calculate_hrv(rr_intervals)
            
            # Detect arrhythmias (enhanced rule-based + ML features)
            classification, abnormalities, risk_level = self.detect_arrhythmias(heart_rate, rr_intervals)
            
            # Calculate confidence based on signal quality and data
            if signal_quality['quality'] == 'Good' and len(r_peaks) > 10:
                confidence = 0.85 + (min(len(r_peaks), 50) / 50) * 0.15
            elif signal_quality['quality'] == 'Fair' and len(r_peaks) > 5:
                confidence = 0.65 + (min(len(r_peaks), 50) / 50) * 0.20
            else:
                confidence = 0.50
            
            # Determine rhythm
            if len(rr_intervals) > 0:
                rr_std = np.std(rr_intervals)
                rr_mean = np.mean(rr_intervals)
                cv = (rr_std / rr_mean) * 100 if rr_mean > 0 else 100
                rhythm = "Regular" if cv < 10 else "Irregular"
            else:
                rhythm = "Unknown"
            
            return {
                'classification': classification,
                'confidence': round(confidence, 3),
                'risk_level': risk_level,
                'details': {
                    'heartRate': round(heart_rate, 1),
                    'heartRateRange': {
                        'min': round(60 / np.max(rr_intervals), 1) if len(rr_intervals) > 0 else 0,
                        'max': round(60 / np.min(rr_intervals), 1) if len(rr_intervals) > 0 else 0
                    },
                    'rhythm': rhythm,
                    'qrsCount': int(len(r_peaks)),
                    'hrv': hrv_metrics,
                    'abnormalities': abnormalities,
                    'signalQuality': signal_quality
                },
                'analysis_type': 'hybrid_ml_enhanced',
                'recommendations': self._generate_recommendations(classification, risk_level, heart_rate)
            }
            
        except Exception as e:
            logger.error(f"Error in analysis: {str(e)}", exc_info=True)
            return {
                'classification': 'Analysis Error',
                'confidence': 0,
                'risk_level': 'Unknown',
                'details': {
                    'error': str(e),
                    'heartRate': 0,
                    'rhythm': 'Unknown',
                    'abnormalities': []
                },
                'analysis_type': 'error'
            }
    
    def _assess_signal_quality(self, signal_array):
        """Assess ECG signal quality"""
        signal_std = np.std(signal_array)
        signal_range = np.max(signal_array) - np.min(signal_array)
        
        if signal_std < 10 or signal_range < 50:
            quality = 'Poor'
            score = 0.3
        elif signal_std < 50 or signal_range < 200:
            quality = 'Fair'
            score = 0.6
        else:
            quality = 'Good'
            score = 0.9
        
        return {
            'quality': quality,
            'score': score,
            'std_deviation': round(signal_std, 2),
            'signal_range': round(signal_range, 2)
        }
    
    def _generate_recommendations(self, classification, risk_level, heart_rate):
        """Generate personalized recommendations"""
        recommendations = []
        
        if risk_level == "Critical" or risk_level == "High":
            recommendations.append("⚠️ IMPORTANT: Consult a cardiologist as soon as possible")
            recommendations.append("📱 Consider wearing a continuous heart monitor")
        
        if "Tachycardia" in classification:
            recommendations.append("🧘 Practice relaxation techniques and reduce stress")
            recommendations.append("☕ Limit caffeine and stimulant intake")
            recommendations.append("💊 Review medications with your doctor")
        
        if "Bradycardia" in classification:
            recommendations.append("⚡ Monitor for symptoms like dizziness or fatigue")
            recommendations.append("🏃 Discuss exercise tolerance with your doctor")
        
        if "Fibrillation" in classification:
            recommendations.append("💊 Anticoagulation therapy may be needed - consult doctor")
            recommendations.append("🩺 Regular ECG monitoring recommended")
            recommendations.append("🚫 Avoid alcohol and excessive caffeine")
        
        if classification == "Normal Sinus Rhythm":
            recommendations.append("✅ Heart rhythm is normal")
            recommendations.append("🏃 Continue regular exercise and healthy diet")
            recommendations.append("📊 Monitor periodically for baseline comparison")
        
        return recommendations

File: backend/utils/ecgAnalyzer.js
/**
 * ECG Signal Processing and Analysis Utilities
 * Implements Pan-Tompkins QRS Detection Algorithm
 */

class ECGAnalyzer {
  constructor() {
    this.samplingRate = 250; // Hz
  }

  /**
   * Pan-Tompkins QRS Detection Algorithm
   * Returns R-peak indices in the ECG signal
   */
  detectQRS(signal) {
    // Filter out NaN and Infinity values
    const cleanSignal = signal.filter(val => !isNaN(val) && isFinite(val));
    
    if (cleanSignal.length < 100) {
      console.log('❌ Insufficient valid data points after cleaning');
      return [];
    }
    
    console.log(`\n🔬 Starting QRS Detection:`);
    console.log(`  Input signal length: ${cleanSignal.length} samples (${signal.length - cleanSignal.length} invalid values removed)`);
    console.log(`  Signal range: ${Math.min(...cleanSignal).toFixed(2)} to ${Math.max(...cleanSignal).toFixed(2)} mV`);
    console.log(`  Signal mean: ${(cleanSignal.reduce((a,b) => a+b, 0) / cleanSignal.length).toFixed(2)} mV`);
    
    // Step 1: Bandpass filter (5-15 Hz)
    const filtered = this.bandpassFilter(cleanSignal);
    console.log(`  ✓ Bandpass filtered`);
    
    // Step 2: Derivative (emphasize QRS slope)
    const derivative = this.derivative(filtered);
    console.log(`  ✓ Derivative calculated`);
    
    // Step 3: Squaring (amplify higher frequencies)
    const squared = derivative.map(x => x * x);
    console.log(`  ✓ Signal squared`);
    
    // Step 4: Moving window integration
    const integrated = this.movingWindowIntegration(squared, 30); // 30 samples = 120ms window
    console.log(`  ✓ Moving window integration complete`);
    
    // Step 5: Adaptive thresholding to find R-peaks
    const rPeaks = this.findRPeaks(integrated, cleanSignal);
    
    return rPeaks;
  }

  /**
   * Bandpass filter (5-15 Hz) for QRS enhancement
   */
  bandpassFilter(signal) {
    // Simple moving average lowpass followed by highpass
    const lowpass = this.lowpassFilter(signal, 15);
    const highpass = this.highpassFilter(lowpass, 5);
    return highpass;
  }

  /**
   * Lowpass filter using moving average
   */
  lowpassFilter(signal, cutoffHz) {
    const windowSize = Math.floor(this.samplingRate / cutoffHz);
    const filtered = [];
    
    for (let i = 0; i < signal.length; i++) {
      let sum = 0;
      let count = 0;
      
      for (let j = Math.max(0, i - windowSize); j <= Math.min(signal.length - 1, i + windowSize); j++) {
        sum += signal[j];
        count++;
      }
      
      filtered.push(sum / count);
    }
    
    return filtered;
  }

  /**
   * Highpass filter (signal - lowpass)
   */
  highpassFilter(signal, cutoffHz) {
    const lowpass = this.lowpassFilter(signal, cutoffHz);
    return signal.map((val, i) => val - lowpass[i]);
  }

  /**
   * Derivative filter (emphasizes slopes)
   */
  derivative(signal) {
    const derivative = [];
    
    for (let i = 2; i < signal.length - 2; i++) {
      // 5-point derivative: (1/8T)(-x[n-2] - 2x[n-1] + 2x[n+1] + x[n+2])
      const value = (
        -signal[i - 2] - 2 * signal[i - 1] + 
        2 * signal[i + 1] + signal[i + 2]
      ) / 8;
      derivative.push(value);
    }
    
    return derivative;
  }

  /**
   * Moving window integration
   */
  movingWindowIntegration(signal, windowSize) {
    const integrated = [];
    
    for (let i = 0; i < signal.length; i++) {
      let sum = 0;
      const start = Math.max(0, i - windowSize + 1);
      
      for (let j = start; j <= i; j++) {
        sum += signal[j];
      }
      
      integrated.push(sum / windowSize);
    }
    
    return integrated;
  }

  /**
   * Find R-peaks using adaptive thresholding
   */
  findRPeaks(integrated, originalSignal) {
    const rPeaks = [];
    const threshold = this.calculateAdaptiveThreshold(integrated);
    const minPeakDistance = Math.floor(this.samplingRate * 0.3); // Minimum 300ms between peaks (200 BPM max)
    
    // Debug logging
    console.log(`🔍 R-Peak Detection Debug:`);
    console.log(`  Signal length: ${integrated.length}`);
    console.log(`  Threshold: ${threshold.toFixed(2)}`);
    console.log(`  Min peak distance: ${minPeakDistance}`);
    console.log(`  Integrated signal range: ${Math.min(...integrated).toFixed(2)} to ${Math.max(...integrated).toFixed(2)}`);
    
    let inPeak = false;
    let peakStart = 0;
    let lastPeakIndex = -minPeakDistance;
    
    for (let i = 1; i < integrated.length - 1; i++) {
      if (integrated[i] > threshold && !inPeak && (i - lastPeakIndex) >= minPeakDistance) {
        // Start of potential QRS complex
        inPeak = true;
        peakStart = i;
      } else if (integrated[i] < threshold && inPeak) {
        // End of QRS complex - find max in original signal
        const peakIndex = this.findMaxInRange(originalSignal, peakStart, i);
        rPeaks.push(peakIndex);
        lastPeakIndex = peakIndex;
        inPeak = false;
      }
    }
    
    console.log(`  ✓ Found ${rPeaks.length} R-peaks`);
    
    return rPeaks;
  }

  /**
   * Calculate adaptive threshold (15% of max for better AD8232 detection)
   */
  calculateAdaptiveThreshold(signal) {
    const max = Math.max(...signal);
    const min = Math.min(...signal);
    const range = max - min;
    // Use 15% of range above baseline for maximum sensitivity with noisy AD8232 signals
    return min + (range * 0.15);
  }

  /**
   * Find index of maximum value in range
   */
  findMaxInRange(signal, start, end) {
    let maxIndex = start;
    let maxValue = signal[start];
    
    for (let i = start + 1; i <= end; i++) {
      if (signal[i] > maxValue) {
        maxValue = signal[i];
        maxIndex = i;
      }
    }
    
    return maxIndex;
  }

  /**
   * Calculate R-R intervals from R-peaks
   */
  calculateRRIntervals(rPeaks) {
    const rrIntervals = [];
    
    for (let i = 1; i < rPeaks.length; i++) {
      const interval = (rPeaks[i] - rPeaks[i - 1]) / this.samplingRate * 1000; // in ms
      rrIntervals.push(interval);
    }
    
    return rrIntervals;
  }

  /**
   * Calculate heart rate from R-R intervals
   */
  calculateHeartRate(rrIntervals) {
    if (rrIntervals.length === 0) return 0;
    
    const avgRR = rrIntervals.reduce((a, b) => a + b, 0) / rrIntervals.length;
    return Math.round(60000 / avgRR); // BPM
  }

  /**
   * Calculate Heart Rate Variability (HRV) metrics
   */
  calculateHRV(rrIntervals) {
    if (rrIntervals.length < 2) {
      return { sdnn: 0, rmssd: 0, pnn50: 0 };
    }

    // SDNN - Standard deviation of NN intervals
    const mean = rrIntervals.reduce((a, b) => a + b, 0) / rrIntervals.length;
    const variance = rrIntervals.reduce((sum, rr) => sum + Math.pow(rr - mean, 2), 0) / rrIntervals.length;
    const sdnn = Math.sqrt(variance);

    // RMSSD - Root mean square of successive differences
    let sumSquaredDiff = 0;
    for (let i = 1; i < rrIntervals.length; i++) {
      sumSquaredDiff += Math.pow(rrIntervals[i] - rrIntervals[i - 1], 2);
    }
    const rmssd = Math.sqrt(sumSquaredDiff / (rrIntervals.length - 1));

    // pNN50 - Percentage of successive differences > 50ms
    let nn50Count = 0;
    for (let i = 1; i < rrIntervals.length; i++) {
      if (Math.abs(rrIntervals[i] - rrIntervals[i - 1]) > 50) {
        nn50Count++;
      }
    }
    const pnn50 = (nn50Count / (rrIntervals.length - 1)) * 100;

    return { sdnn, rmssd, pnn50 };
  }

  /**
   * Detect arrhythmias based on R-R intervals
   */
  detectArrhythmias(rrIntervals, heartRate) {
    const arrhythmias = [];

    // Bradycardia (HR < 60 BPM)
    if (heartRate < 60 && heartRate > 0) {
      arrhythmias.push({
        type: 'Bradycardia',
        severity: 'low',
        description: 'Heart rate below 60 BPM'
      });
    }

    // Tachycardia (HR > 100 BPM)
    if (heartRate > 100) {
      arrhythmias.push({
        type: 'Tachycardia',
        severity: heartRate > 120 ? 'high' : 'medium',
        description: `Heart rate above 100 BPM (${heartRate} BPM)`
      });
    }

    // Irregular rhythm (high HRV)
    if (rrIntervals.length >= 2) {
      const hrv = this.calculateHRV(rrIntervals);
      
      if (hrv.sdnn > 100) {
        arrhythmias.push({
          type: 'Irregular Rhythm',
          severity: 'medium',
          description: 'High heart rate variability detected'
        });
      }
    }

    // Check for very irregular beats (possible AFib)
    const irregularBeats = this.detectIrregularBeats(rrIntervals);
    if (irregularBeats > 0.3) { // More than 30% irregular
      arrhythmias.push({
        type: 'Possible Atrial Fibrillation',
        severity: 'high',
        description: 'Highly irregular rhythm detected'
      });
    }

    return arrhythmias;
  }

  /**
   * Detect irregular beats (ratio of irregular RR intervals)
   */
  detectIrregularBeats(rrIntervals) {
    if (rrIntervals.length < 3) return 0;

    let irregularCount = 0;
    const mean = rrIntervals.reduce((a, b) => a + b, 0) / rrIntervals.length;

    for (let i = 0; i < rrIntervals.length; i++) {
      // If interval deviates more than 20% from mean, consider irregular
      if (Math.abs(rrIntervals[i] - mean) > mean * 0.2) {
        irregularCount++;
      }
    }

    return irregularCount / rrIntervals.length;
  }

  /**
   * Analyze complete ECG session
   */
  analyzeSession(ecgData) {
    try {
      // Extract voltage values
      const signal = ecgData.map(point => point.voltage_mv || point.voltage || 0);

      // Detect QRS complexes
      const rPeaks = this.detectQRS(signal);

      // Calculate R-R intervals
      const rrIntervals = this.calculateRRIntervals(rPeaks);

      // Calculate heart rate
      const heartRate = this.calculateHeartRate(rrIntervals);

      // Calculate HRV
      const hrv = this.calculateHRV(rrIntervals);

      // Detect arrhythmias
      const arrhythmias = this.detectArrhythmias(rrIntervals, heartRate);

      // Calculate signal quality
      const signalQuality = this.assessSignalQuality(signal);

      return {
        success: true,
        metrics: {
          heartRate,
          rPeakCount: rPeaks.length,
          avgRRInterval: rrIntervals.length > 0 
            ? rrIntervals.reduce((a, b) => a + b, 0) / rrIntervals.length 
            : 0,
          hrv,
          signalQuality
        },
        arrhythmias,
        rPeaks,
        rrIntervals
      };
    } catch (error) {
      console.error('Error analyzing ECG session:', error);
      return {
        success: false,
        error: error.message
      };
    }
  }

  /**
   * Assess signal quality
   */
  assessSignalQuality(signal) {
    // Calculate signal-to-noise ratio (simplified)
    const mean = signal.reduce((a, b) => a + b, 0) / signal.length;
    const variance = signal.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / signal.length;
    const stdDev = Math.sqrt(variance);

    // Check for flat line (very low variance)
    if (stdDev < 10) {
      return { score: 0, status: 'Poor - Flat signal' };
    }

    // Check for excessive noise (very high variance)
    if (stdDev > 500) {
      return { score: 30, status: 'Poor - Noisy signal' };
    }

    // Good signal
    const score = Math.min(100, Math.max(0, 100 - (Math.abs(stdDev - 100) / 5)));
    
    let status;
    if (score > 80) status = 'Excellent';
    else if (score > 60) status = 'Good';
    else if (score > 40) status = 'Fair';
    else status = 'Poor';

    return { score: Math.round(score), status };
  }
}

module.exports = new ECGAnalyzer();
